healpy/rotator.py

```diff
@@ -373,9 +377,44 @@ def angle_ref(self, *args, **kwds):
         cosalpha = north_pole[2] - vp[2] * np.dot(north_pole, vp)
         return np.arctan2(sinalpha, cosalpha)
 
-    def rotate_map(self, m):
-        """Rotate a HEALPix map to a new reference frame
+    def rotate_alm(self, alm, lmax=None, mmax=None):
+        """Rotate Alm to a new reference frame
+
+        see hp.rotate_alm, this function **returns** the rotated alms,
+        does not rotate in place"""
+
+        rotated_alm = alm.copy()  # rotate_alm works inplace
+        rotate_alm(rotated_alm, matrix=self.mat, lmax=lmax, mmax=mmax)
+        return rotated_alm
+
+    def rotate_map_alms(self, m, use_pixel_weights=True, lmax=None, mmax=None):
+        """Rotate a HEALPix map to a new reference frame in spherical harmonics space
+
+        Parameters
+        ----------
+        m : np.ndarray
+            Input map, 1 map is considered I, 2 maps:[Q,U], 3 maps:[I,Q,U]
+        use_pixel_weights : bool, default True
+            Use pixel weights in map2alm
+
+        Returns
+        -------
+        m_rotated : np.ndarray
+            Map in the new reference frame
+        """
+        alm = sphtfunc.map2alm(
+            m, use_pixel_weights=use_pixel_weights, lmax=lmax, mmax=mmax
+        )
+        rotated_alm = self.rotate_alm(alm, lmax=lmax, mmax=mmax)
+        return sphtfunc.alm2map(
+            rotated_alm, lmax=lmax, mmax=mmax, nside=pixelfunc.get_nside(m)
+        )
+
+    def rotate_map_pixel(self, m):
+        """Rotate a HEALPix map to a new reference frame in pixel space
 
+        It is always better to rotate in spherical harmonics space, see
```

Comment:
Well, strictly speaking no. If you have a non-bandlimited map, both methods will give bad results, and at least the pixel based method will be faster ;)

Reply:
what about:

        It is generally better to rotate in spherical harmonics space, see
        the rotate_map_alms method. A case where pixel space rotation is                                                            
        better is for heavily masked maps where the spherical harmonics
        transform is not well defined.

Reply:
It's fine. I'm just worried by the fact that spherical harmonic analysis is not well defined for practically any map in the world, and nobody seems to really care.

I've heard so many people say: "I've run a map through map2alm and then through alm2map and the result is really different from the input." This is _normal_ on the sphere.

Imagine you have a map containing point sources. If you rotate it using rotate_map_alms, you get tons of ringing artifacts ... probably not what you want. And increasing resolution and lmax won't help either.

Reply:
The docstrings should be targeted to the less experienced users, so we should give suggestions that work most of the time. An expert user should know how to go beyond that independently.
I think also mentioning point sources would be useful.

```
    It is generally better to rotate in spherical harmonics space, see
    the rotate_map_alms method. A case where pixel space rotation might be                                                          
    more suiltable is for heavily masked maps where the spherical harmonics
    transform is not well defined or maps with point sources where the spherical harmonics transform introduces ringing artifacts.
```